app/notebook-store.js

```diff
@@ -2,133 +2,53 @@
-        fs.stat(nbpath, function(err, stats) {
-            if (err) {
-                if (err.code === 'ENOENT') { // file not found
-                    resolve(false);
-                } else {
-                    reject(err);
-                }
-            } else {
-                resolve(stats.isFile());
-            }
-        });
-    });
-}
+var _cache = {};
 
-// stat the path in the data directory
-function stat(nbpath) {
-    nbpath = path.join(DATA_DIR, nbpath);
-    return new Promise(function(resolve, reject) {
-        fs.stat(nbpath, function(err, stats) {
-            if (err) {
-                return reject(err);
-            }
-
-            stats.fullpath = nbpath;
-
-            if (stats.isDirectory()) {
-                // check if this directory contains an index dashboard
-                exists(path.join(nbpath, INDEX_NB_NAME)).then(
-                    function success(exists) {
-                        stats.isDashboard = stats.hasIndex = exists;
-                        if (stats.hasIndex) {
-                            // check if bundled dashboard has an 'urth_components' dir
-                            fs.stat(path.join(nbpath, 'urth_components'),
-                                function(err, urth_stats) {
-                                    stats.supportsDeclWidgets = !err && urth_stats.isDirectory();
-                                    resolve(stats);
-                                });
-                        } else {
-                            resolve(stats);
-                        }
-                    },
-                    function failure(err) {
-                        reject(err);
-                    }
-                );
-            } else {
-                // might be a dashboard file or a "regular" file
-                stats.isDashboard = stats.isFile() && path.extname(nbpath) === DB_EXT;
-                resolve(stats);
-            }
-        });
-    });
-}
+//////////////////
+// READ OPERATIONS
+//////////////////
 
 // Read notebook from cache or from file
 function _loadNb(nbpath, stats) {
     if (!stats) {
-        stats = stat(nbpath);
+        stats = nbfs.stat(nbpath);
     }
     return Promise.resolve(stats)
         .then(function(stats) {
-            var nbfile;
-            if (stats.hasIndex) {
-                nbfile = path.join(nbpath, INDEX_NB_NAME);
-            } else {
-                nbfile = _appendExt(nbpath);
-            }
-
+            var nbfile = stats.fullpath ||
+                path.join(DATA_DIR, appendExt(nbpath, DB_EXT));
             return new Promise(function(resolve, reject) {
-                var nbAbsPath = path.join(DATA_DIR, nbfile);
-                fs.readFile(nbAbsPath, 'utf8', function(err, rawData) {
+                fs.readFile(nbfile, ENCODING, function(err, rawData) {
                     if (err) {
-                        reject(new Error('Error loading notebook: ' + err.message));
+                        reject(err);
```

Comment:
Not sure... it looks like the old pattern is still being used below. I'll revert the error to the more descriptive.